### PR TITLE
#686

### LINKED ISSUE #686: Allow different policy length in ParallelogramOLF
### Discussed in https://github.com/casact/chainladder-python/discussions/684

<div type='discussions-op-text'>

<sup>Originally posted by **andyhufford-HARC** March 25, 2026</sup>
Is there documentation for the ParallelogramOLF the takes into account the length of the policy term (i.e. semi-annual vs. annual)? 

approximation_grain was not described in the API Reference, but that appears to be monthly or daily grain noted from another forum (https://github.com/casact/chainladder-python/discussions/497). Is there another parameter for policy-term length that is an input?

`class chainladder.ParallelogramOLF(rate_history=None, change_col='', date_col='', approximation_grain='M', vertical_line=False)`</div>

### Is your feature request aligned with the scope of the package?

- [x] Yes, absolutely!
- [ ] No, but it's still worth discussing.
- [ ] N/A (this request is not a codebase enhancement).

### Describe the solution you'd like, or your current workaround.

_No response_

### Do you have any additional supporting notes?

_No response_

### PR DESCRIPTION
**Pull request by kennethshsu**
## Summary of Changes  
Introduced `policy_length` as a parameter to denote the length of the policy in the parallelogram method, allowing anything in months granularity.

## Related GitHub Issue(s)  
Closes #686 

## Additional Context for Reviewers  
This is a bit hard to review, as it is one of those cases where you need to build an example in Excel or work through it manually to reconcile the package calculation. However, I was able to reconcile it against another example using Excel, which is attached. Please review the calculations in excel too to make sure I didn't do anything weird.

[New Example.xlsx](https://github.com/user-attachments/files/27612159/New.Example.xlsx)

- [x] I passed tests locally for both code (`uv run pytest`) and documentation changes (`uv run jb build docs --builder=custom --custom-builder=doctest`)

---

> [!NOTE]
> **Medium Risk**
> Changes the core `parallelogram_olf` calculation by making the averaging window depend on a new `policy_length` parameter and adjusting the lookback/leap-year handling, which can change on-level factors for existing users. API signature changes (`date`→`dates`) may break downstream calls if they used keyword arguments.
> 
> **Overview**
> Adds a new `policy_length` (months) parameter to `ParallelogramOLF`/`parallelogram_olf` and updates the OLF computation to use policy-length-based rolling windows and lookback periods, with revised daily vs monthly and leap-year handling.
> 
> Updates the estimator to pass `policy_length` through and renames the `parallelogram_olf` argument from `date` to `dates`. Moves and significantly expands test coverage into `test_parallelogram.py` (including `policy_length`, vertical vs non-vertical line, and daily/monthly behavior), removing the older utility-level tests; remaining changes are mostly formatting/typing cleanup.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 088b017504c41c5b609691132a47ebfdc07048e7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by henrydingliu** [comment]
since most of test_policy_length is based on werner modlin, would it be better as an example/doctest?

**Comment by kennethshsu** [comment]
> since most of test_policy_length is based on werner modlin, would it be better as an example/doctest?

Ya, I plan on working on the examples in doctoring and will use the right ticket for it next.

**Comment by henrydingliu** [comment]
gotcha. 

any thoughts on the bugbot reports?

**Comment by kennethshsu** [comment]
> gotcha.
> 
> any thoughts on the bugbot reports?

I took a look at this, and a few thoughts:

1. I think there’s a bug somewhere, but the issue is not as direct as it appears.
2. I tried letting bug bot fix it, but the way it did it is super weird. I’m not convinced it actually resolves the underlying issue (or let alone introduces more issues)
3. I tested another very simple example, and I got the answers I expected. :o
```python
rate_history = pd.DataFrame(
    {
        "EffDate": ["2010-07-01"],
        "RateChange": [0.35],
    }
)

data = pd.DataFrame(
    {"Year": [2010, 2011, 2012, 2013, 2014], "EarnedPremium": [10_000] * 5}
)

prem_tri = cl.Triangle(
    data,
    origin="Year",
    columns="EarnedPremium",
    cumulative=True,
)

prem_tri = cl.ParallelogramOLF(
    rate_history,
    change_col="RateChange",
    date_col="EffDate",
    policy_length=24, # <--- 2 year policy
    approximation_grain="M",
).fit_transform(prem_tri)

prem_tri.olf_
```

There's still probably a bug hiding somewhere in an edge case though...

**Comment by henrydingliu** [comment]
do we have any tests that have non-12month policy and D approximation grain?

**Comment by kennethshsu** [comment]
I have tested 6 month policies so we should be ok there. `approximation_grain` isn't new, so I'm not too worry about that. It is very hard to do by hand/excel though. Do you want to check with a trivial example or something?

**Comment by henrydingliu** [comment]
i only see 6-month tests with monthly grain. and we only had daily tests for 12 month length. i don't think we can automatically assume that daily with non-12 will work.

**Comment by henrydingliu** [comment]
```python
data = [
    [2002, 61183, 0],
    [2003, 69175, .05],
    [2004, 99322, .075],
    [2005, 138151, .15],
    [2006, 107578, .1],
    [2007, 62438, -.2],
    [2008, 47797, -.2]
]
columns = [
    'Calendar Year',
    'Earned Premiums',
    'Rate Changes'
]
df_prem = pd.DataFrame(data, columns=columns)
df_prem['Date'] = pd.to_datetime(df_prem['Calendar Year'].astype(int).astype(str) + '-01-01')
comparison = cl.parallelogram_olf(df_prem['Rate Changes'],df_prem['Date']).reset_index()
comparison['pol_12'] = cl.parallelogram_olf(df_prem['Rate Changes'],df_prem['Date'],policy_length = 12).reset_index()['OLF']
comparison['pol_6'] = cl.parallelogram_olf(df_prem['Rate Changes'],df_prem['Date'],policy_length = 6).reset_index()['OLF']
comparison['pol_6_daily'] = cl.parallelogram_olf(df_prem['Rate Changes'],df_prem['Date'],policy_length = 6, approximation_grain = 'D').reset_index()['OLF']
print(comparison)
```
```python
  Origin       OLF    pol_12     pol_6  pol_6_daily
0   2002  0.913836  0.913836  0.913836          NaN
1   2003  0.891547  0.891547  0.880806          NaN
2   2004  0.838863  0.838863  0.823972     0.809677
3   2005  0.753116  0.753116  0.727730          NaN
4   2006  0.670476  0.670476  0.654884          NaN
5   2007  0.711111  0.711111  0.752941          NaN
6   2008  0.888889  0.888889  0.941176     0.999659
```

**Comment by kennethshsu** [comment]
@henrydingliu I figured out the bug where it returned `NaN`, but I don't actually know if the calculated OLFs are correct lol... And your example is too hard to do by hand x_x

File: chainladder/adjustments/tests/test_parallelogram.py [hunk 1]
```diff
@@ -32,3 +34,252 @@ def test_parallelogram(clrd):
...
+    )
+    assert (
+        cl.parallelogram_olf(
+            df_prem["Rate Changes"],
+            df_prem["Date"],
+            policy_length=6,
+            approximation_grain="D",
+        )
+        .reset_index()["OLF"]
+        .notna()
```
Comment (henrydingliu) [inline chainladder/adjustments/tests/test_parallelogram.py:283]: i think we can put together a stronger test here. monthly approximation and daily approximation should return identical results when rate changes are only happening at the beginning of the year.  rather than `notna`, can we do (monthly result - daily result) within some tolerance?

Comment (kennethshsu) [reply chainladder/adjustments/tests/test_parallelogram.py:283]: I like where you are going with this. I added two more tests: 

If rate is effective 1/1, there's no difference in monthly or daily approximation.
If rate is effective 7/1, the monthly will have higher OLFs because there are less days in the first half of the year.

What else do you got? To be honest I think with these simple examples we can work out the actual factors by hand.

**Review by henrydingliu — DISMISSED** [review]
didn't looked through all the linting changes in utility_function.py. bugbot didn't notice any functional changes.

File: chainladder/adjustments/parallelogram.py [hunk 2]
```diff
@@ -21,10 +21,19 @@ class ParallelogramOLF(BaseEstimator, TransformerMixin, EstimatorIO):
         5% decrease should be stated as -0.05
     date_col: str
         A list-like set of effective dates corresponding to each of the changes
-    approximation_grain: str
-        The resolution of the internal calendar used for calculating the on-level factors: 
-        monthly ('M') or daily ('D'). Daily is finer and adjusts for leap years when assigning
-        factors to origin periods.
+    approximation_grain: str {"M", "D"} (default="M")
```
[the remark this answers is not in the document]
Comment (henrydingliu) [reply chainladder/adjustments/parallelogram.py:24]: seems like a minor issue. i'm good with this PR if you are

[the remark this answers is not in the document]
Comment (kennethshsu) [reply chainladder/adjustments/parallelogram.py:24]: Just saw it, I'll fix it

**Review by henrydingliu — APPROVED** [review]
